client/core/core.go

```diff
@@ -0,0 +1,251 @@
+		c.connMtx.RUnlock()
+	}
+}
+
+// addDex adds a dexConnection to the conns map if a connection can be made
+// and the DEX configuration is successfully retrieved. The connection is
+// unauthenticated until the `connect` request is sent and accepted by the
+// server.
+func (c *Core) addDex(uri string) {
+	// Get the host from the DEX URL.
+	parsedURL, err := url.Parse(uri)
+	if err != nil {
+		log.Errorf("error parsing account URL %s: %v", uri, err)
+		return
+	}
+	// Create a websocket connection to the server.
+	conn, err := comms.NewWsConn(&comms.WsCfg{
+		URL:      "wss://" + parsedURL.Host + "/ws",
+		PingWait: 60 * time.Second,
+		RpcCert:  c.certs[uri],
+		ReconnectSync: func() {
+			go c.handleReconnect(uri)
+		},
+		Ctx: c.ctx,
+	})
+	if err != nil {
+		log.Errorf("Error creating websocket connection for %s: %v", uri, err)
+		return
+	}
+	// Request the market configuration. The DEX is only added when the DEX
+	// configuration is successfully retrieved.
+	reqMsg, err := msgjson.NewRequest(conn.NextID(), msgjson.ConfigRoute, nil)
+	if err != nil {
+		log.Errorf("error creating 'config' request: %v", err)
+	}
+	conn.Request(reqMsg, func(msg *msgjson.Message) {
+		resp, err := msg.Response()
+		if err != nil {
+			log.Errorf("failed to parse 'config' response message: %v", err)
+			return
+		}
+		var dexCfg *msgjson.ConfigResult
+		err = json.Unmarshal(resp.Result, dexCfg)
+		if err != nil {
+			log.Errorf("failed to parse config response")
+			return
+		}
+		assets := make(map[uint32]*msgjson.Asset)
+		for _, asset := range dexCfg.Assets {
+			assets[asset.ID] = &asset
+		}
+		// Validate the markets so we don't have to check every time later.
+		for _, bq := range dexCfg.Markets {
+			_, ok := assets[bq[0]]
+			if !ok {
+				log.Errorf("%s reported a market with base asset %d, but did not provide the asset info.", uri, bq[0])
+			}
+			_, ok = assets[bq[1]]
+			if !ok {
+				log.Errorf("%s reported a market with quote asset %d, but did not provide the asset info.", uri, bq[1])
+			}
+		}
+		// Create the dexConnection and add it to the map.
+		dc := &dexConnection{
+			conn:   conn,
+			assets: assets,
+			cfg:    dexCfg,
+		}
+		c.connMtx.Lock()
+		c.conns[uri] = dc
+		c.connMtx.Unlock()
+		c.wg.Add(1)
+		// Listen for incoming messages.
+		go c.listen(dc)
+	})
+}
+
+// handleReconnect is called when a WsConn indicates that a lost connection has
+// been re-established.
+func (c *Core) handleReconnect(uri string) {
+	log.Infof("DEX at %s has reconnected", uri)
+}
+
+// listen monitors the DEX webscocket connection for server requests and
+// notifications.
+func (c *Core) listen(dc *dexConnection) {
+	msgs := dc.conn.MessageSource()
+	defer c.wg.Done()
+out:
+	for {
+		select {
+		case msg := <-msgs:
+			switch msg.Route {
+			// Requests
+			case msgjson.MatchDataRoute:
+				log.Info("match_data message received")
+			case msgjson.MatchProofRoute:
```

Comment:
I know these are placeholders, but maybe switch on `Type` too/first to show responses.